cranelift-codegen/src/regalloc/coloring.rs

```diff
@@ -256,6 +271,46 @@ impl<'a> Context<'a> {
 
         for (lv, abi) in args.iter().zip(&sig.params) {
             match lv.affinity {
+                Affinity::RegUnit(unit) => {
+                    if let ArgumentLoc::Reg(reg) = abi.location {
+                        if unit == reg {
+                            if !lv.is_dead {
+                                let bank = self.reginfo.bank_containing_regunit(reg).unwrap();
+                                let rc = self.reginfo.classes
+                                    [bank.first_toprc..(bank.first_toprc + bank.num_toprcs)]
+                                    .iter()
+                                    .find(|&rc| rc.contains(unit))
+                                    .expect("reg unit should be in a toprc");
+                                regs.take(rc, reg, lv.is_local);
+                            }
+                            self.cur.func.locations[lv.value] = ValueLoc::Reg(reg);
+                        } else {
+                            // TODO: what does this mean? A value is assigned somewhere by the ABI
```

Comment:
So this is a really weird one to me. If `unit != reg`, my understanding is that the ABI-assigned location disagrees with where we think the value should be (so if a value comes in as a function argument, passed to an internal call in some other location?). In that case, I initially deferred to the declared Affinity, but that resulted in invalid register constraints. As-is, this uses whatever the ABI-assigned location is, and that seems to do things right, but that's still exactly opposite of what I expected.

Reply:
This is a good question, and I don't have an answer right now. Does this happen as a result of the code not being able to move ABI arguements into specified `RegUnit` affinities?

Reply:
That's my understanding, though the cause is interesting. After making [`parameter affinity more granular`](https://github.com/CraneStation/cranelift/pull/782/commits/4511b31a230998ec05002970128c09b86cf7ac8d) I brought back this [`debug_assert!`](https://github.com/CraneStation/cranelift/pull/782/commits/59603adc5d401f078c22ba36af5649d09dabfc92#diff-ac2d83406d2fa759e3c3ffa71ae7d551R264). It now passes all tests, including the `spill.clif` filetest, which was the failure before.

My understanding is that a value going into a call in `spill` was in `r10` that ends up on the stack by ABI, while I was overwriting another value's affinity to also be `r10`. I think two values having the same affinity is what resulted in the invalid constraints.